### PR TITLE
feat(orchestration): phase-aware Baton composition (v1.2.0)

### PR DESCRIPTION
**Pull request by Nanako0129**
## What changed

Prepare pilotfish v1.2.0 as a phase-aware orchestration release instead of a one-size-fits-all dispatch brake.

| Area | Change |
|---|---|
| Discovery | Stabilize question, scope, evidence format, and stop condition; permit only bounded read-only work that reduces Plan uncertainty |
| Plan and approval | Keep synthesis in the main session; add tool-enforced read-only `plan-verifier` (`READY` / `REVISE`) and require explicit approval before writes |
| Execution | Preserve the strict stable-scope, exclusive-ownership, done-criteria, integration, and verification brake |
| Verification | Keep outcome `verifier` read-and-run after implementation (`CONFIRMED` / `REFUTED`); separate it from the pre-approval role |
| Model routing | Existing named roles continue to omit invocation-level `model`; frontmatter remains the sole model source |
| Security | Add tool-enforced read-only `security-reviewer` before approval; write-capable `security-executor` accepts only approved implementation contracts |
| Long processes | Remove detach-and-yield across all four Bash-capable leaf roles; hand back command plus absolute worktree/cwd, environment, and inputs for tracked main-session execution |
| Runtime safety | Require Claude Code 2.1.207+, the verified baseline that enforces the pre-approval roles' `tools` allowlists; older or unidentifiable builds stop before writes |
| Evidence | Publish the complete bilingual Baton Gate, exact prompts, rejected harness run, exact tested policy/agents snapshot, normalized routing evidence, hashes, metrics, and machine-readable results |

The release also preserves the original dispatch experiment and re-scopes its early Baton probes as incomplete discovery-only observations rather than a compatibility failure.

## Why

The earlier brake conflated two different questions. Discovery needs a stable research contract but may happen before the implementation outcome is known; writing agents need a stable execution contract and required approval. Treating both phases alike either suppresses useful evidence gathering or dispatches executors before the main session can own a coherent Plan.

A dual-mode verifier prompt was insufficient because Plan readiness happens before approval but the outcome verifier needs Bash after implementation. The final design separates capability: `plan-verifier` has only Read/Glob/Grep, while outcome `verifier` retains Bash but no write tools. Security analysis and implementation use the same reviewer/executor split.

Separately, four direct harness trials from @dromsak showed that `nohup` / `setsid` escape task tracking, while a timeout-promoted command under a foreground-spawned agent is terminated shortly after the agent returns. The minimal prose fix from `fix-subagent-detach` is included without adopting PR #11's plugin/hook replatform.

This was initially prepared under a v1.1.6 working stamp, then reclassified as v1.2.0 before release because it changes orchestration behavior and adds roles and capability boundaries rather than delivering a patch-only fix.

## Live Gate

The corrected fresh-session Gate used native first-party Claude routing with the installed Baton skill and the PR candidate policy.

| Turn | Wall time | Client-reported cost | API turns | Result |
|---|---:|---:|---:|---|
| Discovery + Plan | 221.661 s | $1.763515 | 18 | Baton loaded; direct discovery; zero writes; read-only Opus `plan-verifier` used only Glob/Read and returned `READY` |
| Approved execution + verification | 226.487 s | $2.025533 | 4 | Sonnet `mech-executor` wrote only `REPORT.md`; `npm test` passed; fresh Opus outcome `verifier` returned `CONFIRMED` |
| Total | 448.148 s | $3.789048 | 22 | Full lifecycle passed without resend |

All three named-role calls omitted invocation-level `model`. Baton chose direct main-session Discovery, then delegated the stable approved writing contract to `mech-executor`; this validates phase-dependent topology rather than mandatory fan-out.

The exact runtime policy and eight-role `--agents` JSON are committed under `benchmarks/baton-compatibility/final-gate-snapshot/`. That evidence retains the v1.1.6 candidate stamp used during the run. The v1.2.0 release policy differs only in the inert version comment; `results.json` records both raw hashes and tests lock the normalized policies together.

The first isolation attempt is disclosed but rejected: excluding the user setting source hid the installed Baton skill, so no approval turn was run. An earlier complete dual-mode-verifier Gate is also preserved but superseded after review found its no-write boundary was prompt-only. Raw transcript hashes, snapshots, and exact prompts remain under `benchmarks/baton-compatibility/`.

The final fixture did not trigger long work or invoke `security-reviewer`. Long-process evidence remains the contributor's four dedicated direct trials plus regression coverage across every Bash-capable role; the security reviewer boundary is its positive tool allowlist plus policy tests. No extra edge-case Claude run is claimed as coverage.

## Other controls

| Control | Observed result |
|---|---|
| Stable 12-file edit | Opus 4.8 main session → `mech-executor`; 12/12 pass; in the execution-only segment, reported cost field down 36.01% with wall time up 7.92%; neither comparison included the required outcome verifier, so this proves route reachability rather than full-lifecycle savings |
| Small task-local audit | Direct pass beat two scouts by 11.71% wall time and 15.61% reported cost in one comparison |
| Tightly coupled bug | Main session retained diagnosis/fix; fresh verification remained intact |

These are single-run observations. Client-reported cost is not a provider invoice and no population-performance claim is made.

## Checks

```text
python3 -m unittest discover -s tests -v
13 tests passed

all repository JSON parsed successfully
all changed Markdown relative links resolved
published historical and release-candidate hashes matched
privacy scan passed
research fixture npm test passed
fresh shallow-clone pinned-worktree replay passed
git diff --check passed
```

This PR is ready for review. It does not create the v1.2.0 tag or release.

### CODE REVIEW

**Comment by dromsak** [comment]
Flagging a separate finding that surfaced while I was working in this area — it's **orthogonal to the dispatch brake** (this PR answers *whether delegation is worth it*; the finding below is a bug in *how the long-running-process handoff behaves*), so it doesn't conflict with the direction here. But it touches the same policy file, and it means the current detach-and-yield rule is actively harmful, so I wanted it in front of you before v1.1.6.

### The finding

The rule in `executor`/`mech-executor` ("launch it detached with `nohup` + log, end your turn") and the orchestrator-side rule I added in #7 ("arm a background wait on the yielded PID") are both compensating for a harness behaviour that, tested directly, works differently than either rule assumes.

**A foreground Bash command that exceeds its `timeout` is not killed — it's promoted to a background task** ("Command running in background with ID: … You will be notified when it completes"). Whether that promise is kept depends on how the orchestrator spawned the agent:

| Subagent spawned with | Promoted command's fate |
|---|---|
| `run_in_background: true` | **Survives.** Runs to completion, stdout captured throughout, completion notification re-invokes the agent. Observed surviving 59s / 69s / 81s past a turn boundary and **141s past the agent returning its final result** — zero signals, across 4 trials. |
| `run_in_background: false` (foreground) | **`SIGTERM`ed ~4s after the agent returns.** Work destroyed, captured output truncated mid-stream. Reproduced exactly: heartbeat ends at `SIGNAL SIGTERM`, task output file truncated at the same tick. |

`nohup`/`setsid` dodge that `SIGTERM` by escaping the process group — which is exactly why the pattern looks like it works — but they **also escape the harness's task tracking**: no task id, no captured output, no completion notification. So detaching doesn't rescue the handoff; it converts a *destroyed* result into an *orphaned* one, and #7 exists to hand-roll a recovery for the tracking that `nohup` threw away.

I chased three wrong mechanisms before this held up (killed-at-turn-boundary, SIGPIPE-on-stdout, stranded-but-uncollected) — each reproduced cleanly and then failed a control. The decisive variable turned out to be foreground-vs-background *spawn*, nothing in the command itself.

### The correction

- No role is told to detach. Subagents run in the foreground with an explicit `timeout` and hand back anything that can't finish in one.
- **Long-running processes belong to the orchestrator** — the only context whose background tasks are both tracked and reliably notified.
- This makes `run_in_background: true` on agent spawns **load-bearing for correctness, not just cost** — worth stating so it isn't "optimized" back to a foreground spawn to grab a result inline. (That intersects your net-benefit wording here: a foreground spawn to wait idle isn't only wasteful, for a long command it's destructive.)

This supersedes my own #7, which the evidence now shows was patching a self-inflicted wound.

### If useful

- Minimal v1.x-shaped fix (no version bump), matching the current architecture: [`dromsak:fix-subagent-detach`](https://github.com/dromsak/pilotfish/tree/fix-subagent-detach) — replaces the detach instruction in the two executors and the #7 bullet, plus a regression test.
- The same rule enforced by a `PreToolUse` hook rather than requested in prose (subagents can't `run_in_background` or `nohup` at all; the main session keeps both): #11.

Happy to fold the prose fix into this PR instead if you'd rather keep v1.1.6 as one changeset — your call on sequencing. Raw traces/fixtures available if you want to reproduce the SIGTERM.

**Comment by Nanako0129** [comment]
Integrated the minimal long-process correction in ff9f118: executor roles no longer detach, the main orchestrator owns tracked background commands, and a regression test locks the boundary. The v1.1.6 changelog credits @dromsak and the commit carries a Co-authored-by trailer. PR #11's plugin/hook replatform remains separate.

File: templates/claude-md.orchestration.md [hunk 1]
```diff
@@ -1,33 +1,42 @@
...
+- **Long-running processes are yours, not a subagent's.** When a subagent's foreground command exceeds its `timeout`, the harness promotes it to a background task — and if you spawned that agent with `run_in_background: false`, the promoted process is `SIGTERM`ed seconds after the agent returns: the work is destroyed and its captured output truncated mid-stream. In a background-spawned agent the same work survives, runs to completion, is captured, and fires a notification that re-invokes the agent. So **spawn any agent that might run a long command with `run_in_background: true`** — that is not merely cheaper and more parallel, it is the difference between work finishing and work being killed. Subagents therefore never detach anything (`nohup`/`setsid` escape the harness's task tracking entirely, which is how a handoff gets orphaned); when one reports that its task needs a long-running process, that is a correct handoff — run the command yourself with `Bash(run_in_background: true)`, the only context whose background tasks are both tracked and reliably notified, then resume the agent with the output.
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply templates/claude-md.orchestration.md]: Updated in 16b9650. Long-command handoffs now require the exact command, absolute working directory (including the isolated worktree), required environment variables, and input paths. The orchestration contract requires main-session Bash to run in that exact context rather than the parent checkout, and the policy regression test locks those fields.

File: benchmarks/dispatch-brake/positive-controls/README.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  /path/to/pilotfish/benchmarks/dispatch-brake/positive-controls/mechanical/task.md \
+  | sed '1d;$d')"
+
+/usr/bin/time -p claude -p "$TASK" \
+  --output-format stream-json \
+  --verbose \
+  --no-session-persistence \
+  --dangerously-skip-permissions \
+  --max-budget-usd 3 \
+  --append-system-prompt-file /path/to/pilotfish/templates/claude-md.orchestration.md
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply benchmarks/dispatch-brake/positive-controls/README.md]: Updated in 16b9650. The reproduction now mounts the pinned 863b117 snapshot, builds JSON from all six role definitions in that snapshot, passes it explicitly through --agents, and injects the policy from the same snapshot. The current checkout supplies only the generic JSON builder. I also exercised the documented snapshot/build path locally and confirmed six roles were produced.

File: benchmarks/dispatch-brake/positive-controls/README.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+|---|---|---:|---:|---|
+| pilotfish mechanical, hard veto | Inline | 128.24 s | $0.790263 | 12/12 pass |
+| pilotfish mechanical, balanced | `mech-executor` foreground | 138.40 s | $0.505682 | 12/12 pass |
+| pilotfish small research, broad fan-out | 2 scouts background | 261.52 s | $1.036893 | Pass |
+| pilotfish small research, direct comparison | Inline | 234.10 s | $0.896864 | Pass |
+| pilotfish small research, sized gate | Inline | 228.96 s | $0.918431 | Pass |
+| pilotfish tightly coupled bug, balanced | Inline | 77.45 s | $0.365309 | 2/2 pass |
+| remora tightly coupled bug, balanced | Inline diagnosis/fix → verifier foreground | 200.86 s | $0.817504 | 2/2 pass |
+
+For the mechanical control, delegation reduced the reported cost field by 36.01% while adding 7.92% wall time. For the small research control, the two-scout run increased wall time by 11.71% and the reported cost field by 15.61% relative to its direct comparison. These are single task-local runs, not population estimates, and the research comparison did not include downstream Plan synthesis or execution.
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply benchmarks/dispatch-brake/positive-controls/README.md]: Scoped in 16b9650 without claiming an unmeasured lifecycle result. The table and bilingual prose now state that both mechanical runs omitted outcome verification and that 36.01% is execution-only evidence of route reachability, not full-lifecycle savings. Both result JSON files encode outcome_verification_included=false and full_lifecycle_savings_claimed=false.

File: benchmarks/baton-compatibility/results.json [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    "commit": "5f027b8c7701d9acd4ae424a6089ffe3d1fa57a7",
+    "path": "benchmarks/dispatch-brake/positive-controls/research/fixture"
+  },
+  "runtime": {
+    "claude_code": "2.1.207",
+    "calico_patched_binary": true,
+    "provider_route": "native Claude first-party authentication",
+    "candidate_base_head": "863b117b9da42179c5bb77a05158920fbc092ee2",
+    "gate_orchestration_sha256": "b2854ae1a32875144108c69f68d1a75ec4c755e935bbd9fb25214a95c449b0f7",
+    "gate_agents_json_sha256": "9d9b4f80ce8aa1f7f3e1386655a991dee9025726f2882ae0ac9cd2a32f788b46",
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply benchmarks/baton-compatibility/results.json]: Published in 9b7e4d6. benchmarks/baton-compatibility/gate-snapshot/ now contains the exact project CLAUDE.md and the exact session-scoped agents.json used by the accepted Gate. The stored policy bytes hash to the recorded gate policy value; shell command substitution of agents.json hashes to the recorded agents value. results.json points to both files, a regression test locks both hashes, and the snapshot README marks these historical inputs as evidence-only because later long-process fixes supersede them.

File: benchmarks/baton-compatibility/README.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+  "$SOURCE/templates/agents")"
+cd "$WORK"
+```
+
+The user setting source is intentional: Baton was installed under the user skill directory. Excluding `user` makes the Skill tool report `Unknown skill`. The project-level candidate policy is more specific than user memory, and session-scoped `--agents` definitions take precedence over user agent files.
+
+```bash
+claude --dangerously-skip-permissions \
+  -p --output-format json --max-budget-usd 3 \
+  --session-id SESSION_ID --model best --effort high \
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply benchmarks/baton-compatibility/README.md]: Updated in 9b7e4d6. The reproduction now generates SESSION_ID with Python uuid.uuid4() and quotes the same value in both --session-id and --resume. Regression assertions lock the generation and both uses; I also generated and parsed a UUID through this exact command locally.

File: benchmarks/dispatch-brake/positive-controls/README.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+npm test
+
+TASK="$(sed -n '/^```text$/,/^```$/p' \
+  "$SNAPSHOT/benchmarks/dispatch-brake/positive-controls/mechanical/task.md" \
+  | sed '1d;$d')"
+AGENTS_JSON="$(python3 \
+  "$HARNESS/benchmarks/baton-compatibility/build-agents-json.py" \
+  "$SNAPSHOT/templates/agents")"
+
+/usr/bin/time -p claude -p "$TASK" \
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply benchmarks/dispatch-brake/positive-controls/README.md:74]: Updated in 9b7e4d6. The mechanical reproduction now passes --model claude-opus-4-8, matching the main-session model recorded for mechanical-pilotfish-balanced. The surrounding claim is also narrowed to byte-identical policy and role inputs; generated output, timing, and client-reported cost remain explicitly non-deterministic single-run observations.

File: templates/claude-md.orchestration.md [hunk 7]
```diff
@@ -1,33 +1,42 @@
...
 - Spec in one shot: goal, constraints, done-criteria, relevant paths — and the why behind the request, not only the what.
 - Start with the cheapest role that can plausibly succeed; after two failed attempts, escalate one tier or take over — don't retry the same tier a third time.
+- Route every security-sensitive task (authn/authz, secrets, crypto, validation, hardening, vulnerability analysis) to `security-executor`, never the main session or a general executor. Before required approval, give it only an explicit `ANALYSIS ONLY` brief that forbids edits; after approval, give it the stable implementation contract.
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply templates/claude-md.orchestration.md]: Resolved in 40b9b7f with a capability split. New security-reviewer uses the positive allowlist Read, Glob, Grep, WebSearch, WebFetch and owns pre-approval evidence only. The write-capable security-executor now rejects briefs without an approved stable execution contract. Policy and regression tests lock that routing; the report explicitly discloses that the security path was statically verified rather than adding an unrelated runtime task.

File: templates/claude-md.orchestration.md [hunk 8]
```diff
@@ -1,33 +1,42 @@
...
+- **Long-running processes are yours, not a subagent's.** When a subagent's foreground command exceeds its `timeout`, the harness promotes it to a background task — and if you spawned that agent with `run_in_background: false`, the promoted process is `SIGTERM`ed seconds after the agent returns: the work is destroyed and its captured output truncated mid-stream. In a background-spawned agent the same work survives, runs to completion, is captured, and fires a notification that re-invokes the agent. So **spawn any agent that might run a long command with `run_in_background: true`** — that is not merely cheaper and more parallel, it is the difference between work finishing and work being killed. Subagents therefore never detach anything (`nohup`/`setsid` escape the harness's task tracking entirely, which is how a handoff gets orphaned). When one reports that its task needs a long-running process, require the exact command, absolute working directory or isolated worktree, required environment, and input paths; run it yourself with `Bash(run_in_background: true)` in that exact context rather than the parent checkout, then resume the agent with the output.
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply templates/claude-md.orchestration.md]: Resolved in 40b9b7f. The no-detach, ten-minute foreground ceiling, and exact command/cwd/worktree/environment/input handoff now appear directly in every Bash-capable leaf prompt: mech-executor, executor, verifier, and security-executor. The policy names the same four-role set and the regression test iterates over all four. The existing four direct harness trials remain the runtime evidence; this small Baton fixture did not claim to trigger a long process.

File: templates/claude-md.orchestration.md [hunk 9]
```diff
@@ -1,33 +1,42 @@
...
 - Model routing is owned by agent definitions. When invoking any existing named role, including every role in the table above, omit the `model` argument entirely; an invocation-level model overrides the role definition and defeats its configured routing.
 - Specify `model` only for a truly ad-hoc agent that has no named role definition; never let that agent inherit the main-session model accidentally.
-- Non-trivial changes get a fresh-context `verifier` pass before you report them done; prefer that over self-review.
+- Every `verifier` invocation must name exactly one mode and its matching verdict vocabulary. A Plan-readiness brief requests only **READY** / **REVISE** and must not request **CONFIRMED** / **REFUTED**; an outcome-verification brief requests only **CONFIRMED** / **REFUTED** and must not request **READY** / **REVISE**.
+- Material Plans may get a fresh-context `verifier` readiness pass before approval; non-trivial completed changes get a fresh-context `verifier` pass before you report them done. Prefer independent refutation over self-review, while keeping final judgment and synthesis in the main session.
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply templates/claude-md.orchestration.md]: Resolved in 40b9b7f by separating Plan and outcome capabilities. New plan-verifier has only Read, Glob, and Grep, so Bash and every write tool are absent before approval; outcome verifier retains Bash after implementation and returns only CONFIRMED/REFUTED. The final native-Claude two-turn Gate was rerun against the exact eight-role candidate: plan-verifier actually used only Glob/Read, returned READY with a clean Git tree, then approved execution and a fresh outcome verifier completed with CONFIRMED. Exact inputs, hashes, metrics, and the superseded dual-mode Gate are published.

File: README.md [hunk 10]
```diff
@@ -61,34 +61,50 @@ flowchart TD
...
 
-The six roles:
+The eight roles:
 
 | Role | Model | Effort | Used for |
 |---|---|---|---|
 | `scout` | haiku | low | Read-only lookups: "where/how is X", symbol usages, config values |
 | `Explore` | haiku | low | Overrides the built-in Explore agent (see warning above) |
+| `plan-verifier` | opus | medium | Tool-enforced read-only Plan challenge before approval; returns READY/REVISE |
+| `security-reviewer` | opus | high | Tool-enforced read-only security evidence and threat review before approval |
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply README.md:84]: Fixed in 1f1e76f. Both README uninstall prompts now remove all eight installed pilotfish agent files, and the policy test locks the documented count.

File: README.md [hunk 11]
```diff
@@ -182,10 +198,11 @@ The delegation policy in `CLAUDE.md` speaks only of roles (`executor`, `scout`,
...
-| Doesn't spawning agents cost extra? | Yes — every spawn is a fresh context that re-reads its slice of the codebase, and spec-writing costs main-session tokens. That overhead is why the policy says don't delegate single-file reads or quick judgments. The savings come from volume work (search, bulk edits, test runs), where the cheaper tier's per-token price dwarfs the spawn overhead. |
+| I also installed a delegation-planning skill | Treat it as a complementary planning layer. A skill such as [Baton](https://github.com/cablate/baton) can shape discovery questions, worker count, ownership, sequence, and stop conditions; pilotfish supplies the named Claude roles, model routing, leaf-agent boundary, approval gate, and verifier contract. The [public two-turn compatibility gate](./benchmarks/baton-compatibility/README.md) completed Discovery → main-session Plan → `READY` → explicit approval → execution → fresh `CONFIRMED` verification. On its small fixture Baton correctly chose direct Discovery and writing rather than spawning workers. pilotfish never disables user skills. |
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply README.md]: Fixed in 1f1e76f. Both FAQ translations now match the recorded Gate: Discovery stayed in the main session, the approved mechanical write went to mech-executor, and verifier performed the fresh outcome check.

File: templates/agents/plan-verifier.md [hunk 12]
```diff
@@ -0,0 +1,18 @@
+---
+name: plan-verifier
+description: Read-only fresh-context adversarial review of a material Plan before approval. Give it the Plan and evidence paths; it challenges assumptions, scope, ownership, sequencing, stop conditions, and acceptance checks, then returns READY or REVISE. It never executes commands, writes the Plan, edits files, or fixes implementation.
+model: opus
+effort: medium
+tools: Read, Glob, Grep
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply templates/agents/plan-verifier.md:6]: Fixed in 1f1e76f. The runbook now gates installation on Claude Code 2.1.207 or newer, the verified tool-allowlist-enforcing baseline, and stops before any write plan or mutation when the runtime is older or unidentifiable. Both READMEs state the same hard requirement, and a regression test locks the gate.

File: benchmarks/dispatch-brake/positive-controls/README.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+
+## Reproduce
+
+To replay the published balanced mechanical harness with byte-identical policy and role inputs, attach the pinned `863b117` snapshot as a temporary worktree. The current checkout supplies only the generic JSON builder; both the policy and all six role definitions come from the pinned snapshot and are injected explicitly, so no global pilotfish install is required. The main session is pinned to the recorded Opus 4.8 model; generated output, timing, and cost remain single-run observations rather than deterministic bytes.
+
+```bash
+HARNESS=/path/to/current/pilotfish
+SNAPSHOT=/tmp/pilotfish-dispatch-863b117
+
+git -C "$HARNESS" worktree add --detach "$SNAPSHOT" 863b117
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply benchmarks/dispatch-brake/positive-controls/README.md]: Fixed in 1ac02e7. Both replay guides now fetch the full pinned SHA before adding the worktree, and a regression test locks the ordering. I also validated the exact sequence from a fresh --depth 1 clone: the repository remained shallow and the detached worktree resolved to 863b117b9da42179c5bb77a05158920fbc092ee2.
